packages/playground/src/services.ts

```diff
@@ -0,0 +1,188 @@
+
+  const serverHost: ServerHost = {
+    compilerHost: host,
+    getDocumentByURL(url: string) {
+      const model = monaco.editor.getModel(monaco.Uri.parse(url));
+      if (model) {
+        return TextDocument.create(url, "cadl", 1, model.getValue());
+      }
+
+      return undefined;
+    },
+    sendDiagnostics() {},
+    // eslint-disable-next-line no-console
+    log: console.log,
+  };
+  const serverLib = createServer(serverHost);
+  function textDocumentForModel(model: monaco.editor.IModel) {
+    return TextDocument.create(model.uri.toString(), "cadl", 1, model.getValue());
+  }
+
+  monaco.languages.registerDefinitionProvider("cadl", {
+    async provideDefinition(model, position) {
+      const result = await serverLib.gotoDefinition({
+        position: {
+          line: position.lineNumber - 1,
+          character: position.column - 1,
+        },
+        textDocument: textDocumentForModel(model),
+      });
+
+      if (!result) return;
+
+      const loc: monaco.languages.Location = {
+        uri: monaco.Uri.parse(result.uri),
+        range: {
+          startColumn: result.range.start.character + 1,
+          startLineNumber: result.range.start.line + 1,
+          endColumn: result.range.end.character + 1,
+          endLineNumber: result.range.end.line + 1,
+        },
+      };
+
+      return loc;
+    },
+  });
+
+  const lsConfig = serverLib.initialize({
+    capabilities: {},
+    processId: 1,
+    workspaceFolders: [],
+    rootUri: "inmemory://",
+  });
+  serverLib.initialized(lsConfig);
+
+  monaco.languages.registerCompletionItemProvider("cadl", {
+    triggerCharacters: lsConfig.capabilities.completionProvider!.triggerCharacters,
+    async provideCompletionItems(model, position) {
+      const result = await serverLib.complete({
+        position: {
+          line: position.lineNumber - 1,
+          character: position.column - 1,
+        },
+        textDocument: textDocumentForModel(model),
+      });
+      const word = model.getWordUntilPosition(position);
+      const range = {
+        startLineNumber: position.lineNumber,
+        endLineNumber: position.lineNumber,
+        startColumn: word.startColumn,
+        endColumn: word.endColumn,
+      };
+
+      const suggestions: monaco.languages.CompletionItem[] = [];
+      for (const item of result.items) {
+        suggestions.push({
+          label: item.label,
+          kind: item.kind as any,
+          documentation: item.documentation,
+          insertText: item.insertText!,
+          range,
+          commitCharacters:
+            item.commitCharacters ?? lsConfig.capabilities.completionProvider!.allCommitCharacters,
+        });
+      }
+
+      return { suggestions };
+    },
+  });
+
+  const tokenTypes = [
+    "comment",
+    "string",
+    "number",
+    "keyword",
+    "namespace",
+    "variable",
+    "type",
+    "function",
+    "operator",
+    "source",
+  ];
+
+  function mapToken(tok: Token) {
```

Comment:
Cosider filing an issue to implement semantic highligiting in language service and to reuse its classification here. This might be the first reason that I'm inclined to bother with semantic highlighting in LSP. Until now, it didn't seem worth it since it is not recommended to only have that without tmlanguage.

Reply:
I can file this too.

Reply:
This isn't good I don't think. We likely need a parse tree to give better highlighting than the tmlanguage, or at least some clever management of tokenizer state. This just blindly maps tokens to colors which in general doesn't work at all for identifiers which should be colorized differently depending on the syntax they're embedded in.

Reply:
Right, I was saying we should do the good thing, use in other editors and here, that's the issue I was going to file :)

Reply:
https://github.com/microsoft/cadl/issues/310